File: src/lib/polymarket.ts
export interface PolymarketEvent {
  id: string;
  title: string;
  slug: string;
  outcomes: string[];
  outcomePrices: string[];
  volume: number;
  endDate: string;
  active: boolean;
}

export async function fetchAIPolymarketEvents(): Promise<PolymarketEvent[]> {
  try {
    // Polymarket CLOB API — search for AI-related markets
    const res = await fetch(
      "https://gamma-api.polymarket.com/events?tag=ai&limit=10&active=true&closed=false",
      { cache: "no-store" }
    );

    if (!res.ok) return [];

    const AI_KEYWORDS = [
      "ai", "artificial intelligence", "gpt", "claude", "gemini", "llm",
      "openai", "anthropic", "deepmind", "chatgpt", "machine learning",
      "neural", "model", "agi", "robot", "autonomous", "deep learning",
    ];

    const events = await res.json();
    return (events || [])
      .filter((e: Record<string, unknown>) => {
        const title = ((e.title as string) || "").toLowerCase();
        return AI_KEYWORDS.some((kw) => title.includes(kw));
      })
      .map((e: Record<string, unknown>) => ({
        id: e.id as string,
        title: e.title as string,
        slug: e.slug as string,
        outcomes: (e.outcomes as string[]) || [],
        outcomePrices: (e.outcomePrices as string[]) || [],
        volume: (e.volume as number) || 0,
        endDate: e.endDate as string,
        active: e.active as boolean,
      }));
  } catch {
    return [];
  }
}

File: server/app/clients/polymarket.py
import httpx

from app.schemas.news import PolymarketEvent

AI_KEYWORDS = [
    "ai", "artificial intelligence", "gpt", "claude", "gemini", "llm",
    "openai", "anthropic", "deepmind", "chatgpt", "machine learning",
    "neural", "model", "agi", "robot", "autonomous", "deep learning",
]


async def fetch_ai_polymarket_events() -> list[PolymarketEvent]:
    url = "https://gamma-api.polymarket.com/events?tag=ai&limit=10&active=true&closed=false"
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            res = await client.get(url)
        if not res.is_success:
            return []
        events = res.json() or []
        return [
            PolymarketEvent(
                id=str(e.get("id", "")),
                title=e.get("title", ""),
                outcomes=e.get("outcomes") or [],
                outcome_prices=e.get("outcomePrices") or [],
                volume=float(e.get("volume") or 0),
            )
            for e in events
            if any(kw in (e.get("title") or "").lower() for kw in AI_KEYWORDS)
        ]
    except Exception:
        return []
